DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/zero_copy_data_transfer/src/zero_copy_data_transfer.cpp

```diff
@@ -0,0 +1,184 @@
+#include <assert.h>
+#include <algorithm>
+#include <chrono>
+#include <iomanip>
+#include <random>
+#include <type_traits>
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/INTEL/fpga_extensions.hpp>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+#include "buffer_kernel.hpp"
+#include "restricted_usm_kernel.hpp"
+
+using namespace sycl;
+
+// the data type - assert that it is arithmetic
+// this allows the design to easily switch between types by changing
+// the line below
+using Type = int;
+static_assert(std::is_arithmetic<Type>::value);
+
+int main(int argc, char* argv[]) {
+  // parse command line arguments
+#if defined(FPGA_EMULATOR)
+  size_t size = 10000;
+  size_t iterations = 1;
+#else
+  size_t size = 100000000;
+  size_t iterations = 5;
+#endif
+
+  // Allow the size to be changed by a command line argument
+  if (argc > 1) {
+    size = std::stoi(std::string(argv[1]));
+  }
+
+  // check the size
+  if (size <= 0) {
+    std::cerr << "ERROR: size must be greater than 0\n";
+    return 1;
+  }
+
+  try {
+    // device selector
+#if defined(FPGA_EMULATOR)
+    INTEL::fpga_emulator_selector selector;
+#else
+    INTEL::fpga_selector selector;
+#endif
+
+    // queue properties to enable profiling
+    auto prop_list = property_list{ property::queue::enable_profiling() };
+
+    // create the device queue
+    queue q(selector, dpc_common::exception_handler, prop_list);
+
+    // make sure the device supports USM host allocations
+    device d = q.get_device();
+    if (!d.get_info<info::device::usm_host_allocations>()) {
+      std::cerr << "ERROR: The selected device does not support USM host"
+                << " allocations\n";
+      return 1;
+    }
+
+    // the golden output
+    std::vector<Type> out_gold(size);
+
+    // input and output data for the buffer version
+    std::vector<Type> in_buffer(size), out_buffer(size);
+
+    // input and output data for the restricted USM version
+    // malloc_host allocates memory specifically in the host's address space
+    Type* in_restricted_usm = malloc_host<Type>(size, q.get_context());
```

Comment:
Similar comment on malloc_host.  Why ctx instead of just q?

For sycl::free you don't need the prefix since you are not using it elsewhere.